### PR TITLE
Prototype: abstract Table/Query objects, Analysis operations [work in progress]

### PR DESCRIPTION
**Pull request by andy-esch**
This PR is a proof of concept of Table operations as they could be implemented in cartoframes. Read the docs for a fuller picture: https://cartoframes.readthedocs.io/en/prototype-table-objects/cartoframes.analysis.html#module-cartoframes.analysis

cartoframes_table_objects

Beyond these hard-coded methods, we should allow for a generic node. 

```python
t = Table('brooklyn_poverty')
t.intersects('thing1', how='spatial').node({'def': 'operation...'})
```

Ideas:

+ create a read only version that keeps nesting the queries and only reads the result instead of doing write operations on the table
+ Table methods: `t.describe()` shows basic stats on the table similar to `df.describe()` in pandas
+ Chaining works well with the Batch SQL API since you can pass a list of queries to it, and this has the advantage that it won't time out like the SQL API will
+ Look at Spark DataFrame / Spark SQL operations for must haves: http://spark.apache.org/docs/2.1.0/api/python/pyspark.sql.html#pyspark.sql.DataFrame
+ Scikit-learn's Pipeline constructs: http://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html
+ `lazy=True|False` to control how it's evaluated? Or.. like dask, do a `.compute()` to initiate the run?
+ Add a `table.sql('...')` method which has syntax for referencing the previous node
+ Augment points to data observatory measure

## Functions to implement

See a brain-stormed list here: https://github.com/CartoDB/cartoframes/issues/254 . Also checkout spark's sql function module: http://spark.apache.org/docs/2.2.0/api/python/pyspark.sql.html#module-pyspark.sql.functions

closes https://github.com/CartoDB/cartoframes/issues/12

cc @stuartlynn

### CODE REVIEW

**Comment by alrocar** [comment]
Stale issue. Closing 👋 

The idea still seems useful to me though

**Comment by andy-esch** [comment]
No problem, but I want to keep the branch because I have some documentation built off of it
